doc/regexp.rdoc

```diff
@@ -1,753 +1,1176 @@
+- +m+: <tt>/_pattern_/m</tt> sets
+  {Multiline Mode}[rdoc-ref:regexp.rdoc@Multiline+Mode].
+- +x+: <tt>/_pattern_/x</tt> sets
+  {Extended Mode}[rdoc-ref:regexp.rdoc@Extended+Mode].
+- +o+: <tt>/_pattern_/o</tt> sets
+  {Interpolation Mode}[rdoc-ref:regexp.rdoc@Interpolation+Mode].
+
+Any, all, or none of these may be applied.
+
+Modifiers +i+, +m+, and +x+ may be applied to subexpressions:
+
+- A preceding <tt>(?_modifier_)</tt> turns the mode "on."
+- A preceding <tt>(?-_modifier_)</tt> turns the mode "off."
+
+Example:
+
+  re = /(?i)te(?-i)st/
+  re.match('test') # => #<MatchData "test">
+  re.match('TEst') # => #<MatchData "TEst">
+  re.match('TEST') # => nil
+  re.match('teST') # => nil
+
+\Method Regexp#options returns an integer whose value showing
+the settings for case-insensitivity mode, multiline mode, and extended mode.
+
+=== Case-Insensitive Mode
+
+By default, a regexp is case-sensitive:
+
+  /foo/.match('FOO')  # => nil
+
+Modifier +i+ enables case-insensitive mode:
+
+  /foo/i.match('FOO')
+  # => #<MatchData "FOO">
+
+\Method Regexp#casefold? returns whether the mode is case-insensitive.
+
+=== Multiline Mode
+
+By default, a Ruby regexp is in multiline mode,
+which means that anchors <tt>^</tt> and <tt>$</tt> match, respectively,
+the beginning or end of a line in the target string.
+
+Modifier +m+ therefore does not enable multiline mode -- it's already enabled.
+The modifier does, however, make one small difference:
+
+- Without the +m+ modifier, a newline does not match subexpression <tt>.</tt>:
+
+    /a.c/.match("a\nc")  # => nil
+
+- With the modifier, it does match:
+
+    /a.c/m.match("a\nc") # => #<MatchData "a\nc">
+
+=== Extended Mode
+
+Modifier +x+ enables extended mode, which means that:
+
+- Literal white space in the pattern is to be ignored.
+- Character <tt>#</tt> marks the remainder of its containing line as a comment,
+  which is also to be ignored.
+
+In extended mode, whitespace and comments may be used
+to form a self-documented regexp.
+
+Regexp not in extended mode (matches some Roman numerals):
+
+  pattern = '^M{0,3}(CM|CD|D?C{0,3})(XC|XL|L?X{0,3})(IX|IV|V?I{0,3})$'
+  re = /#{pattern}/
+  re.match('MCMXLIII') # => #<MatchData "MCMXLIII" 1:"CM" 2:"XL" 3:"III">
+
+Regexp in extended mode:
+
+  pattern = <<-EOT
+    ^                   # beginning of string
+    M{0,3}              # thousands - 0 to 3 Ms
+    (CM|CD|D?C{0,3})    # hundreds - 900 (CM), 400 (CD), 0-300 (0 to 3 Cs),
+                        #            or 500-800 (D, followed by 0 to 3 Cs)
+    (XC|XL|L?X{0,3})    # tens - 90 (XC), 40 (XL), 0-30 (0 to 3 Xs),
+                        #        or 50-80 (L, followed by 0 to 3 Xs)
+    (IX|IV|V?I{0,3})    # ones - 9 (IX), 4 (IV), 0-3 (0 to 3 Is),
+                        #        or 5-8 (V, followed by 0 to 3 Is)
+    $                   # end of string
+  EOT
+  re = /#{pattern}/x
+  re.match('MCMXLIII') # => #<MatchData "MCMXLIII" 1:"CM" 2:"XL" 3:"III">
+
+=== Interpolation Mode
```

Comment:
The `o` flag means that the first time the literal regexp with interpolations is encountered, the generated Regexp object is saved and used for all future evaluations of the literal regexp.  Without the `o` flag, the generated Regexp is not saved, so each evaluation of the literal regexp generates a new Regexp object.  Note that if the literal regexp does not have interpolations, the `o` behavior is the default.

I recommend this section be rewritten with that in mind.

Reply:
This section still needs to be rewritten IMO.

Reply:
Rewritten.